### ISSUE TITLE
optimization: adding impossible condition to a complex query results in Fixpoint error

### ISSUE THREAD

**Issue by asenac** [MaterializeInc/materialize#8391]
### What version of Materialize are you using?

```
$ materialized -v
materialized v0.9.5-dev (f500d4e13) + the fix for #8387  
```

### How did you install Materialize?

* [ ] Docker image
* [ ] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [x] Built from source
* [ ] Materialize Cloud

### What was the issue?

Adding a `WHERE FALSE` clause to the complex query in #8387, leads to the following fix point issue:

```
ERROR:  internal transform error: fixpoint looped too many times Fixpoint {
    transforms: [
        FuseAndCollapse {
            transforms: [
                ProjectionExtraction,
                ProjectionLifting,
                Map,
                Negate,
                Filter,
                Project,
                Join,
                InlineLet {
                    inline_mfp: false,
                },
                Reduce,
                Union,
                UnionBranchCancellation,
                UpdateLet,
                RedundantJoin,
                FoldConstants {
                    limit: Some(
                        10000,
                    ),
                },
            ],
        },
    ],
    limit: 100,
} %0 = Let l0 =
| Constant

%1 =
| Get %0 (l0)
| Distinct group=(#0, #1)
| Negate

%2 =
| Constant (null)

%3 =
| Join %1 %2
| | implementation = Unimplemented

%4 = Let l1 =
| Union %0 %3

%5 = Let l2 =
| Constant

%6 =
| Get ? (u1)

%7 = Let l3 =
| Join %5 %6
| | implementation = Unimplemented
| Filter (#0 = #2), (#1 = #3), ("foo1" = (#4 ->> "type")), (1632340042327 <= (1000 * f32tonumeric(date_part_epoch_ts((#5 + 90 days))))), (1632340042327 >= (1000 * f32tonumeric(date_part_epoch_ts(#5))))
| Reduce group=(#0, #1)
| | agg max(#5)

%8 =
| Get %7 (l3)
| Distinct group=(#0, #1)
| Negate

%9 =
| Get %5 (l2)
| Distinct group=(#0, #1)

%10 =
| Union %8 %9

%11 =
| Constant (null)

%12 =
| Join %10 %11
| | implementation = Unimplemented

%13 = Let l4 =
| Union %7 %12

%14 =
| Get %13 (l4)
| Reduce group=(#0, #1)
| | agg count(true)
| Filter (#2 > 1)
| Map (err: more than one record produced in subquery)
| Project (#0, #1, #3)

%15 = Let l5 =
| Union %13 %14

%16 = Let l6 =
| Constant

%17 = Let l7 =
| Get %16 (l6)
| Distinct group=(#0, #1)

%18 =
| Get ? (u1)

%19 = Let l8 =
| Join %17 %18
| | implementation = Unimplemented
| Filter (#0 = #2), (#1 = #3), ("foo1" = (#4 ->> "type")), (1632340042327 <= (1000 * f32tonumeric(date_part_epoch_ts((#5 + 90 days))))), (1632340042327 >= (1000 * f32tonumeric(date_part_epoch_ts(#5))))
| Reduce group=(#0, #1)
| | agg max(#5)

%20 =
| Get %19 (l8)
| Distinct group=(#0, #1)
| Negate

%21 =
| Get %17 (l7)
| Distinct group=(#0, #1)

%22 =
| Union %20 %21

%23 =
| Constant (null)

%24 =
| Join %22 %23
| | implementation = Unimplemented

%25 = Let l9 =
| Union %19 %24

%26 =
| Get %25 (l9)
| Reduce group=(#0, #1)
| | agg count(true)
| Filter (#2 > 1)
| Map (err: more than one record produced in subquery)
| Project (#0, #1, #3)

%27 = Let l10 =
| Union %25 %26

%28 =
| Get %27 (l10)
| Distinct group=(#0, #1)
| Negate

%29 =
| Get %17 (l7)
| Distinct group=(#0, #1)

%30 =
| Union %28 %29

%31 =
| Constant (null)

%32 =
| Join %30 %31
| | implementation = Unimplemented

%33 =
| Union %27 %32

%34 =
| Join %16 %33 (= #0 #300) (= #1 #301)
| | implementation = Unimplemented
| Project (#5, #8, #11, #14, #17, #20, #23, #26, #29, #32, #35, #38, #41, #44, #47, #50, #53, #56, #59, #62, #65, #68, #71, #74, #77, #80, #83, #86, #89, #92, #95, #98, #101, #104, #107, #110, #113, #116, #119, #122, #125, #128, #131, #134, #137, #140, #143, #146, #149, #152, #155, #158, #161, #164, #167, #170, #173, #176, #179, #182, #185, #188, #191, #194, #197, #200, #203, #206, #209, #212, #215, #218, #221, #224, #227, #230, #233, #236, #239, #242, #245, #248, #251, #254, #257, #260, #263, #266, #269, #272, #275, #278, #281, #284, #287, #290, #293, #296, #299, #302)

%0 = Let l0 =
| Constant

%1 =
| Get %0 (l0)
| Distinct group=(#0, #1)
| Negate

%2 =
| Constant (null)

%3 =
| Join %1 %2
| | implementation = Unimplemented

%4 = Let l1 =
| Union %0 %3

%5 = Let l2 =
| Constant

%6 =
| Get ? (u1)

%7 = Let l3 =
| Join %5 %6
| | implementation = Unimplemented
| Filter (#0 = #2), (#1 = #3), ("foo1" = (#4 ->> "type")), (1632340042327 <= (1000 * f32tonumeric(date_part_epoch_ts((#5 + 90 days))))), (1632340042327 >= (1000 * f32tonumeric(date_part_epoch_ts(#5))))
| Reduce group=(#0, #1)
| | agg max(#5)

%8 =
| Get %7 (l3)
| Distinct group=(#0, #1)
| Negate

%9 =
| Get %5 (l2)
| Distinct group=(#0, #1)

%10 =
| Union %8 %9

%11 =
| Constant (null)

%12 =
| Join %10 %11
| | implementation = Unimplemented

%13 = Let l4 =
| Union %7 %12

%14 =
| Get %13 (l4)
| Reduce group=(#0, #1)
| | agg count(true)
| Filter (#2 > 1)
| Map (err: more than one record produced in subquery)
| Project (#0, #1, #3)

%15 = Let l5 =
| Union %13 %14

%16 =
| Constant

materialize=> 

```

**Comment by asenac** [comment]
The query has 101 subqueries and the fixpoint performs 100 iterations at most. By just removing two subqueries from the query the fixpoint issue is gone. Each iteration removes the machinery for a single subquery during FoldConstants, until InlineLet unblocks FoldConstants again by inlining the first constant Get:

```
} %0 = Let l0 =
| Constant

%1 =
| Get %0 (l0)
| Distinct group=(#0, #1)
| Negate
```

**Comment by asenac** [comment]
This issue was fixed by #8582 where fixpoints became soft limits that keep going if the number of nodes in the plan keeps decreasing.
